doc/developer/design/20230214_transform_using.md

```diff
@@ -0,0 +1,274 @@
+in the future is maintained until the end of its validity period and `Q_mfp` has preserved the
+previously computed timestamps. This technique is similar to how the `UPSERT` envelope recovers its
+state from its output.
+
+### Planner/Optimizer/Data structure stabilization
+
+The inability to have `RAW` snapshots readily available has some deep implications on how much of
+our internal choices we are willing to stabilize, which in turn have deep implications on the kind
+of queries we want to allow users to use in the `TRANSFORM USING` position.
+
+In the equations above we had to define `C_intermediate` to effectively be a collection whose schema
+is very much dependent on planner and optimizer choices. Since the transformation from `RAW` to
+`C_intermediate` is not described by high level `SQL` that can be stored in the catalog we must
+explore what else we must stabilize in order to be able to continue ingesting a source across
+versions.
+
+There are various flavors of how this could go. I will lay out some of them below.
+
+**Option 1: Only accept transforms `Q` whose `Q_expr` is the identity, stored as SQL**
+
+We make the guarantee that if a version of Materialize plans and optimizes a piece of SQL to just an
+MFP then all future versions will continue to do so and will be producing an identical MFP. No data
+structure stabilization needed.
+
+**Option 2: Only accept transforms `Q` whose `Q_expr` is the identity, stored as raw MFP**
+
+We make the guarantee that an MFP produced by some version of Materialize will be able to be read
+and interpreted in the intended way by all future versions. In practice this means stabilizing
+`MirScalarExpr`, `UnaryFunc`, `BinaryFunc`, and `VariadicFunc`.
+
+**Option 3: Accept any transform `Q`, stored as SQL**
+
+We make the guarantee that if a version of Materialize plans and optimizes a piece of SQL to a pair
+of `Q_mfp` and `Q_expr` then all future versions will continue to produce an identical `Q_mfp` but a
+potentially different `Q_expr`. No data structure stabilization needed.
+
+**Option 4: Accept any transform `Q`, stored as `MirRelationExpr`**
+
+We make no guarantees to the properties of the planner/optimizer over time but we guarantee that a
+`MirRelationExpr` produced by a version of Materialize will be readable in the intended way by all
+future versions. This requires stabilizing everything that was needed for option 2 and additionally
+`MirRelationExpr`, `JoinImplementation`, `AggregateExpr` and perhaps more data strucutres that are
+transitively depended on by `MirRelationExpr`.
+
+> Petros’ take: Stabilizing data structures seems easier to think about than stabilizing code
+> behavior so I would shy away from options 1 and 3. Stabilizing `MirScalarExpr` seems tractable
+> given its simplicity so I would go for option 2 unless an amazing option 5 appears. Input from
+> COMPUTE folks would be helpful here.
```

Comment:
Stabilizing `MirScalarExpr` still stabilizes code behavior, though! Because you still have to evaluate that `MirScalarExpr`, and all that code is subject to change. For sure it's a lot less code than "all code that parses, plans, and executes SQL" though.